### PR TITLE
cluster: adding host rewrite

### LINKED ISSUE #16775: feature regression: no longer possible to rewrite authority
*Title*: no longer possible to rewrite authority

*Description*:
https://github.com/envoyproxy/envoy/pull/14747 changed the behavior of header modifications that are allowed. Previously, we relied on this to rewrite the authority on a per-destination basis.

After the change, this is no longer possible. While we can do a rewrite at the route level, we cannot do it on a per-weighted-cluster basis, unlike all other headers.

For us this represents a breaking change, so we will likely set the runtime flag for now. But we want to understand a long term plan.

cc @alyssawilk 

[optional *Relevant Links*:]
Istio specific issue: https://github.com/istio/istio/issues/33226

### PR DESCRIPTION
**Pull request by alyssawilk**
Risk Level: low (config guarded addition)
Testing: new unit tests
Docs Changes: in API docs
Release Notes: inline
Fixes #16775

### CODE REVIEW

File: api/envoy/config/route/v3/route_components.proto [hunk 1]
```diff
@@ -378,6 +378,13 @@ message WeightedCluster {
     // :ref:`FilterConfig<envoy_v3_api_msg_config.route.v3.FilterConfig>`
     // message to specify additional options.]
     map<string, google.protobuf.Any> typed_per_filter_config = 10;
+
+    oneof host_rewrite_specifier {
+      // Indicates that during forwarding, the host header will be swapped with
+      // this value.
+      string host_rewrite_literal = 11
+          [(validate.rules).string = {well_known_regex: HTTP_HEADER_VALUE strict: false}];
+    }
```
Comment (htuch) [inline api/envoy/config/route/v3/route_components.proto:387]: Seems fine, but do we anticipate much more substantial growth here? I take it this is mostly a byproduct of stricter restrictions on what can be set via `response_headers_to_add` etc. Should the rationale for using this be stated more explicitly in docs?

Comment (alyssawilk) [reply api/envoy/config/route/v3/route_components.proto:387]: your call - as with all Envoy config I think we add it as needed.  I think it's reasonable to allow host/authority/scheme rewrites anywhere we allow request/response header mutations, but no need to clutter up configs/code adding them where they're not needed?

Comment (htuch) [reply api/envoy/config/route/v3/route_components.proto:387]: I was chatting with @stevenzzzz about opportunities to save memory on repeated fields in `RouteAction` today. Using hierarchy is one blunt tool for this, so we may want more fields available for setting at `VirtualHost` and then overrriding at different levels. But, we'll need to confront the existing prior art with request/response header modifications, so I think we can land this PR independent of that.

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**